### PR TITLE
safeguard example

### PR DESCRIPTION
**Pull request by FelipeAdachi**
This PR adds a tutorial example for validating and moderating user prompts and LLM responses.

### CODE REVIEW

File: langkit/whylogs/example_utils/guardrails_example_utils.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+def generate_message_id():
+    myuuid = uuid.uuid4()
+    return str(myuuid)
+
+
+def _generate_response(prompt):
+    """This is where we would ask a model to generate a response to a prompt.
+    Let's just find the response in our data for now.
+    """
+    for d in _data:
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_example_utils.py:34]: Maybe make put the responses in a dictionary keyed by the prompt. Then you could `return response[prompt]` without searching.

File: langkit/whylogs/example_utils/guardrails_llm_schema.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    response: str
+
+
+moderation_queue = {}
+
+
+# Toxic Response Validator
+def nontoxic_condition(msg) -> bool:
+    score = toxicity.toxicity(msg)
+    if score > 0.8:
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_llm_schema.py:39]:
```suggestion
    return score <= 0.8
```

File: langkit/whylogs/example_utils/guardrails_llm_schema.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        return False
+    return True
+
+
+def flag_toxic_response(val_name: str, cond_name: str, value: Any, m_id) -> None:
+    message_metadata: MessageMetadata = moderation_queue.get(m_id, {})
+    message_metadata["toxic_response"] = True
+    message_metadata["response"] = value
+    moderation_queue[m_id] = message_metadata
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_llm_schema.py:48]: `global moderation_queue`

File: langkit/whylogs/example_utils/guardrails_llm_schema.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+# Toxic Prompt Validator
+
+
+def flag_toxic_prompt(val_name: str, cond_name: str, value: Any, m_id) -> None:
+    message_metadata: MessageMetadata = moderation_queue.get(m_id, {})
+    message_metadata["toxic_prompt"] = True
+    message_metadata["prompt"] = value
+
+    moderation_queue[m_id] = message_metadata
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_llm_schema.py:68]: `global moderation_queue`

File: langkit/whylogs/example_utils/guardrails_llm_schema.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    name="nontoxic_prompt",
+    conditions=nontoxic_prompt_conditions,
+    actions=[flag_toxic_prompt],
+)
+
+
+# Forbidden Patterns Validator
+def no_patterns_condition(msg) -> bool:
+    pattern = regexes.has_patterns(msg)
+    if pattern:
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_llm_schema.py:84]:
```suggestion
    return bool(pattern)
```

File: langkit/whylogs/example_utils/guardrails_llm_schema.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        return False
+    return True
+
+
+def flag_patterns_response(val_name: str, cond_name: str, value: Any, m_id) -> None:
+    message_metadata: MessageMetadata = moderation_queue.get(m_id, {})
+    message_metadata["patterns_in_response"] = True
+    message_metadata["response"] = value
+
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_llm_schema.py:93]: `global moderation_queue`

File: langkit/whylogs/example_utils/guardrails_llm_schema.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    moderation_queue[m_id] = message_metadata
+
+
+no_patterns_response_conditions = {
+    "no_patterns_response": Condition(Predicate().is_(no_patterns_condition))
+}
+patterns_response_validator = ConditionValidator(
+    name="nontoxic_prompt",
```
Comment (richard-rogers) [inline langkit/whylogs/example_utils/guardrails_llm_schema.py:101]: Probably not `nontoxic_prompt` ?

File: langkit/whylogs/example_utils/guardrails_example_utils.py (hunk 1 above)
Comment (FelipeAdachi) [reply langkit/whylogs/example_utils/guardrails_example_utils.py]: much better. thanks!

File: langkit/whylogs/example_utils/guardrails_llm_schema.py (hunk 2 above)
Comment (FelipeAdachi) [reply langkit/whylogs/example_utils/guardrails_llm_schema.py]: much better!

**Review by jamie256 — APPROVED** [review]

**Pull request merged**